// stylelint-scss/dist/rules/selector-nest-combinators/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (expectation) {
  return function (root, result) {
    var validOptions = _stylelint.utils.validateOptions(result, ruleName, {
      actual: expectation,
      possible: ["always", "never"]
    });

    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      (0, _utils.parseSelector)(rule.selector, result, rule, function (fullSelector) {
        // attribute, class, combinator, comment, id, nesting, pseudo, root, selector, string, tag, or universal
        var chainingTypes = ["attribute", "class", "id", "pseudo", "tag", "universal"];

        fullSelector.walk(function (node) {
          if (expectation === "always") {
            if (node.type === "selector") {
              return;
            }

            if (node.parent && node.parent.type === "selector" && node.parent.parent && node.parent.parent.type === "pseudo") {
              return;
            }

            if (!node.prev()) {
              return;
            }

            if (node.next() && precedesParentSelector(node)) {
              return;
            }

            if (!chainingTypes.includes(node.type)) {
              return;
            }

            if (node.prev().type === "combinator") {
              if (!node.prev().prev()) {
                return;
              }

              if (!chainingTypes.includes(node.prev().prev().type)) {
                return;
              }
            }

            if (node.prev().type !== "combinator" && chainingTypes.includes(node.type) && !chainingTypes.includes(node.prev().type)) {
              return;
            }
          }

          if (expectation === "never") {
            if (rule.parent.type === "root" || rule.parent.type === "atrule") {
              return;
            }
          }

          _stylelint.utils.report({
            ruleName: ruleName,
            result: result,
            node: rule,
            message: messages.rejected,
            index: node.sourceIndex
          });

          function precedesParentSelector(current) {
            do {
              current = current.next();

              if (current.type === "nesting") {
                return true;
              }
            } while (current.next());

            return false;
          }
        });
      });
    });
  };
};

var _stylelint = require("stylelint");

var _utils = require("../../utils");

var ruleName = exports.ruleName = (0, _utils.namespace)("selector-nest-combinators");

var messages = exports.messages = _stylelint.utils.ruleMessages(ruleName, {
  expected: function expected(prop) {
    return "Expected selector \"" + prop + "\" to be in a nested form";
  },
  rejected: function rejected(prop) {
    return "Unexpected \"" + prop + "\" found in selector";
  }
});
